plugin/pkg/global/resourcereferencemanager/admission.go

```diff
@@ -319,6 +322,68 @@ func (r *ReferenceManager) Admit(ctx context.Context, a admission.Attributes, o
 				})
 			}
 		}
+	case core.Kind("CloudProfile"):
+		cloudProfile, ok := a.GetObject().(*core.CloudProfile)
+		if !ok {
+			return apierrors.NewBadRequest("could not convert resource into CloudProfile object")
+		}
+		if utils.SkipVerification(operation, cloudProfile.ObjectMeta) {
+			return nil
+		}
+		if a.GetOperation() == admission.Update {
+			oldCloudProfile, ok := a.GetOldObject().(*core.CloudProfile)
+			if !ok {
+				return apierrors.NewBadRequest("could not convert old resource into CloudProfile object")
+			}
+
+			// getting Kubernetes versions that have been removed from the CloudProfile
+			removedKubernetesVersions := getRemovedVersions(oldCloudProfile.Spec.Kubernetes.Versions, cloudProfile.Spec.Kubernetes.Versions)
+
+			// getting Machine image versions that have been removed from the CloudProfile
+			removedMachineImageVersions := map[string]sets.String{}
+			for _, oldImage := range oldCloudProfile.Spec.MachineImages {
+				imageFound := false
+				for _, newImage := range cloudProfile.Spec.MachineImages {
+					if oldImage.Name == newImage.Name {
+						imageFound = true
+						removedMachineImageVersions[oldImage.Name] = getRemovedVersions(oldImage.Versions, newImage.Versions)
+					}
+				}
+				if !imageFound {
+					for _, version := range oldImage.Versions {
+						if removedMachineImageVersions[oldImage.Name] == nil {
+							removedMachineImageVersions[oldImage.Name] = sets.NewString()
+						}
+						removedMachineImageVersions[oldImage.Name] = removedMachineImageVersions[oldImage.Name].Insert(version.Version)
+					}
+				}
+			}
+
+			if len(removedKubernetesVersions) > 0 || len(removedMachineImageVersions) > 0 {
+				shootList, e := r.shootLister.List(labels.Everything())
+				if e != nil {
+					return err
```

Comment:
❓

Reply:
return BadRequest() instead. The reason why I just returned the error is because this is [how it is done here](https://github.com/gardener/gardener/blob/2c33877a1315fb2d655cb5a3fc93561767a457fe/plugin/pkg/global/resourcereferencemanager/admission.go#L229-L228)